src/package/utils/reflowDOM.ts

```diff
@@ -1,10 +1,41 @@
-export function reflowDOM() {
+const recreateNode = (node: Node) => {
+    if (node.parentNode) {
+        node.parentNode.replaceChild(node.cloneNode(true), node);
+    }
+};
+
+const visit = (node: Node, elementNames?: Array<string> | undefined) => {
+    for (let i = 0; i < node.childNodes.length; i++) {
+        const childNode = node.childNodes[i];
+
+        // re-creates a node in case of:
+        // - no element name filters given
+        // - element name filter matches childNode's element name
+        if (!elementNames || elementNames.indexOf(childNode.nodeName.toUpperCase()) > -1) {
+            console.log('Re-creating DOM node', elementNames);
+            recreateNode(childNode);
+        }
+
+        if (childNode.childNodes) {
+            visit(childNode, elementNames);
+        }
+    }
+};
+
+export function reflowDOM(elementNameOrNames?: Array<string> | string) {
```

Comment:
How events are triggered now:
![image](https://user-images.githubusercontent.com/2901416/64063981-37361880-cbfc-11e9-9140-e27226603949.png)

with old reflow:
![image](https://user-images.githubusercontent.com/2901416/64063984-3dc49000-cbfc-11e9-8b5c-066a4511ea1d.png)


So maybe we should have 1 new function that is used with new system ?
This way we users can choose how they want this to behave, connectedCallback/disconenctedCallback can break stuff in some comonents if they are not checking if its disconnected

Reply:
Well, for me it looks all fine with the order. The execution order is correct, as we can see that the old instances are disconnected first, then for the new instances always the `constructor` is invoked first and then the `connectedCallback`. And thereafter when an HMR takes place, it will be like some new numbers appear (because of the `clone()` happening first of course) with `constructor()` being called. Then probably `disconnectedCallback` on the older number as the first step of replace (`removeChild` internally called) and after that `connectedCallback` on the new number (`appendChild` called internally). 

For sure the synchronous printing in the console cannot reflect the async code execution intuitively but this way of doing things is much more atomic and performant. We can def. implement tests to ensure the order is correct. This is as easy as implementing a Web Component that sets timing flags in a global object variable when `constructor`, `connectedCallback` and `disconnectedCallback` are called. We can `expect` then, that all methods are called and that the time is ascending.

Reply:
@kyr0 
I forgot to tell I stopped it from doing anything to child nodes after it replaced it the top picture.

I did this:
```ts
const visit = (node: Node, elementNames?: Array<string> | undefined) => {
    for (let i = 0; i < node.childNodes.length; i++) {
        const childNode = node.childNodes[i];
        let recreatedNode = false;

        // re-creates a node in case of:
        // - no element name filters given
        // - element name filter matches childNode's element name
        if (!elementNames || elementNames.indexOf(childNode.nodeName.toUpperCase()) > -1) {
            console.log('Re-creating DOM node', elementNames);
            recreateNode(childNode);
            recreatedNode = true;
        }

        if (childNode.childNodes && !recreatedNode) {
            visit(childNode, elementNames);
        }
    }
};
```